SPARK-16478 graphX (added graph caching in strongly connected components)

wesolowskim opened the pull request:
## What changes were proposed in this pull request?

I added caching in every iteration for sccGraph that is returned in strongly connected components. Without this cache strongly connected components returned graph that needed to be computed from scratch when some intermediary caches didn't existed anymore. 
## How was this patch tested?

I tested it by running code similar to the one  [on databrics](https://databricks-prod-cloudfront.cloud.databricks.com/public/4027ec902e239c93eaaa8714f173bcfc/4889410027417133/3634650767364730/3117184429335832/latest.html). Basically I generated large graph  and computed strongly connected components with changed code, than simply run count on vertices and edges. Count after this update takes few seconds instead 20 minutes. 
# statement

contribution is my original work and I license the work to the project under the project's open source license.

Commit:
corrected SPARK-16478 (caching in strongly connected components)

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented:
You have introduced a new problem though -- you're not unpersisting the RDDs you cache, and, you're doing a needless count (minor). Of course it's faster to operate on the final RDD at the end: you did the work as you went instead of at the end. It's actually slower overall though. Right? I'm not yet clear what this is solving

wesolowskim commented:
Let me introduce some data first:
1. SCC run computed on randomly generated graph just like one provided by me on databrics notebook takes about 120s
2. When doing sccGraph.vertices.count on computed graph it takes about 20 min

After caches i proposed scc run is few second longer but count takes few seconds. 

In total original scc and count takes about 22 min (depending on graph generated) and with caches it takes about 2 min 10 seconds. With graph on real world data I work with materializing of returned RDD took so much time I never managed to do it (waiting hours without effect). 

About introducing problems:

1. Original SCC doesn't unpersist workGraph and the problem already exists, but I agree  I added another RDDs that are not unpersisted
2. I tried to unpersist RDDs in original SCC, but it only added computing time (I had to materialize every time reference was going to be lost) but it didn't solve the problem


There is lot's of materialization going on in original SCC implementation (numVetices), and none of it (outside of Pregel) is not unpersisted. When operating on large graph with limited memory and lot's of iterations it causes intermediary (workGraph) RDDs to be removed in LRU fashion. At the same time sccGraph that is returned is not persisted, and when I take some action like count outside of scc after run, there are no RDD's in cache that it depends on. 

I tried few different solutions but only the one I'm proposing worked. 
Check out code I provided on databrics. It should take more then 20 mins to compute. If instead my implementation would be used it would take about less then 3 mins total. I can provide runnable databrics notebook with both solutions.

srowen commented:
You are right, the code already had a `.cache()` call without `.unpersist()`. It eventually gets cleaned up, but not good form and probably didn't actually do much either. Except perhaps break the lineage.

The problem with cache and unpersist is that if nothing materializes the RDD between the two calls, it does nothing. Yes, I know that's why you forced materialization with an extra call to count.

The thing is, caching by itself can't make the underlying computation faster and none of the RDDs that are now cached are used a second time (right?). I think something else is at work here.

One guess is that the current calls to cache(), which don't seem quite right, are the problem. It computes the whole lineage at once and attempts to persist all RDDs at once, competing with others and evicting them and generally wasting time because the RDDs aren't reused at all (I think). If that's true then actually removing all the caching should also help a lot. If not, then, that's not it.

(PS `sccGraphCountVertices` here is superfluous, you can omit it)

wesolowskim commented:
Whole computation is iterative and depends on previous state of sccWorkGraph. It iterates on sccWorkGraph.numVertices which is action under the hood and without caching whole algorithm would be extremely slow, because with every iteration sccWorkGraph  would be computed all over again. Before I came up with this solution I tried to unpersist intermediary sccWorkGraphs, and colleague of mine tried to remove caches. Non of it worked, because of the reasons mentioned above. 
Initially i thought that caching should be removed but there are two reasons that is not the case:

1. Scc algorithm requires knowledge of data (numVertices) that requires action in spark
2. Every iteration is depended on previous state of graph

What is more pregel itself is implemented with caching and immediate materialization. 

If I cannot remove caches (I think I can't) additional ones have to be added. I tested few other solutions and that is the one that turned to be the best in terms of performance. I hope to find an optimal solution.

srowen commented:
I don't think that's true. Each iteration makes a new RDD that depends on the previous iteration's RDD. The execution is actually just a long chain of RDDs if you 'unrolled' it. That's why I think caching actually doesn't do anything here, since no RDD is used more than once. It's also possible I've missed something just reading the code but not seeing it yet.

Still if you're saying it doesn't work well with no caching, then there must be some subtle reason the intermediate RDDs do benefit from caching. I'd probably appeal to the author of this code to weigh in in that case.

wesolowskim commented:
That's how i see it:
First you have workGraph defined and marked to cache:

`var sccWorkGraph = graph.mapVertices { case (vid, _) => (vid, false) }.cache()`

then it is materialized 

`var numVertices = sccWorkGraph.numVertices`

It is also materialized every iteration of both loops:

`while (sccWorkGraph.numVertices > 0 && iter < numIter) {`

and

`while (sccWorkGraph.numVertices < numVertices)`

Now looking inside the loop:
sccWorkGraph is defined upon it's previous "version"

`sccWorkGraph = sccWorkGraph.outerJoinVertices(sccWorkGraph.outDegrees) {
          (vid, data, degreeOpt) => if (degreeOpt.isDefined) data else (vid, true)
        }.outerJoinVertices(sccWorkGraph.inDegrees) {
          (vid, data, degreeOpt) => if (degreeOpt.isDefined) data else (vid, true)
        }.cache()`

Let's assume we don't have cache(). We end up here 
`while (sccWorkGraph.numVertices < numVertices)`

numVertices is `graph.vertices.count()`

With first iteration sccWorkGraph is just computed and we go on. 
With second iteration we have to compute sccWorkGraph that was created in first iteration and then copmute second that is based on the first one and so on. 

It looks to me like that (in brackets there is iteration from which RDD is):
0: sccWorkGraph[0].vertices.count() 
1: sccWorkGraph[0]->sccWorkgraph[1].vertices.count()
2: sccWorkGraph[0]->sccWorkGraph[1]->sccWorkGraph[2].vertices.count()
3:...

If I understand it correctly with each iteration and no cache, count forces to compute whole lineage whereas if we have previous sccWorkgraph cached, only last few stages have to be computed. 

With caches it looks like this
0: sccWorkGraph[0].cache().vertices.count() 
1: sccWorkGraph[0]->sccWorkgraph[1].cache().vertices.count()
2: sccWorkGraph[0]->sccWorkGraph[1]->sccWorkGraph[2].cache().vertices.count()

The same story goes for sccGraph that depends on previous' iteration sccGraph that depends on finalVertices that depends on sccWorkGraph. If I had infinite memory I would have all intermediary sccWorkGraph RDDs cached and materializing sccGraph would be quite quick, but since I don't it is extremely slow. 

The problem is not seen inside run() method directly because sccGraph is not materialized inside. 

I hope that clarifies something. Without some kind of fix current version of strongly connected components is sufficient only for small graphs relative to available memory.

srowen commented:
Oh good point, it does get materialized after `cache()` already because `numVertices` will call `count`. That does mean there's more than one call to evaluate the RDD, and that quite changes things. I see why `sccWorkGraph` works that way.

I'm still a little hazy on `sccGraph` because the same isn't true of it in the current code. I agree that the dependency on `finalVertices` is relevant. As you say, the underlying `sccWorkGraph`s are cached. I think there does end up being a problem because it still means all of the lineage of cached `sccWorkGraph` evaluate at once and try to cache.

I think your change is probably the right thing, then. What we should really do is add `unpersist` calls in the right places for `sccGraph` and `sccWorkGraph`, which is tricky. That would fully optimize this.

(No my point was that the variable `sccGraphCountVertices` itself doesn't do anything. You don't need to store this value. `.count` does something of course.)

wesolowskim commented:
Great we're on the same page finally. I can try to add unpersists however it is indeed tricky and at first try I wasn't able to unpersist every RDD and at the return had few RDD still in cache.

srowen commented:
Yes, elsewhere in graphx I remember trying this and it was tricky because you had to hold a reference to the old RDD, evaluate and persist the next one that depends on it, and _then_ unpersist the old RDD.

Commit:
SPARK-16478 (added unpersist to newly created caches + assure persistence of edges)

wesolowskim commented:
I added unpersists to additionally created caches and checked performance. scc.run is slightly longer, but returned graph is cached (both vertices and edges). 
To fully optimize it and remove all unused graphs I'd have to carefully check everything on large graph since sometimes it's easy to remove cache to early and in this case effect will be seen after some materialization on graph outside scc. 
I hope to get it accepted in current state (that is much more performant at least for larger graphs). 

Later on I'll try to further optimize it(it is tricky to do it elegantly because sccWorkGraph gets overwritten twice before it is materialized, and one loop get's executed at least once).

srowen commented on graphx/src/main/scala/org/apache/spark/graphx/lib/StronglyConnectedComponents.scala:
```diff
@@ -64,11 +69,20 @@ object StronglyConnectedComponents {
-        }
+        }.cache()
+        // take triplet to materialize vertices and edges (faster then count on both)
+        sccGraph.triplets.take(1)
```
It's faster because it only materializes a single partition. I think you have to call something like count.

wesolowskim replied:
I wasn't sure about this take to be honest. I checked both solutions - with take on triplets and with count on vertices and edges separately. After return from scc count on vertices and edges showed it was generally faster with take, but it could be due to small amount of partitions. 
Count on both vertices and edges with randomly generated graph of 500k vertices and 2mln vertices adds about 20s to scc, but may be worth it (didn't show during my tests but I guess you could be right). It could also be that one partition of triplet depends on whole verticesRDD and edgesRDD so, it would be sufficient to make take(1).

srowen commented on graphx/src/main/scala/org/apache/spark/graphx/lib/StronglyConnectedComponents.scala:
```diff
@@ -44,6 +44,11 @@ object StronglyConnectedComponents {
 
+    // helper variables to unpersist cached graphs
+    var prevSccGraph1 = sccGraph
+    var prevSccGraph2 = sccGraph
```
Do you need two vars? seems like they both track the same variable.

wesolowskim replied:
I guess I could avoid to have this two vars, but it was hard to reason about it code with one only. One sccGraph is persisted inside do-while loop, and the second at the end of outside while loop. With two variables it is clearer which graph is unpersisted. For example when do-while loop is left you have to unpersist prevSccGraph1, but not necessarily prevSccGraph2.

srowen replied:
I guess my logic is that it's easier to have one variable following this single variable throughout rather than two -- which of the two refers to which previous RDD? are they different? the effect of this as well is that you unpersist and the re-persist the RDD once during each loop.

wesolowskim replied:
You're totally right. I'll correct it.

srowen commented:
Is it now not necessary to cache the 'work graph' as well?

wesolowskim commented:
'work graph' is already cached and it is indeed necessary. 
To summarize - do you propose to replace sccGraph.triplets.take(1) with sccGraph.vertices.count() and sccGraph.edges.count()? 
Do you agree to leave two vars for previous instance of sccGraph to distinguish where the caching took place?

srowen commented:
Ack right I mean does it need to be explicitly materialized? I suppose the new code also has the effect of materializing the last work RDD, but then I imagine we need to unpersist it too.

wesolowskim commented:
I don't know if I understand you correctly. Last work RDD is explicitly materialized because iteration's depend on number of vertices left in graph. 
My aim with latest solution was to leave algorithm with persisted graph at the and and not to add any additional unpersisted unwanted rdds. I would be nice to have everything cleaned, but since it is tricky I'd like to work on it some time later.

srowen commented:
I was thinking it should be perfectly possible to unpersist scWorkGraph RDDs that are persisted along the way in the same way? it should be the same pattern. It's possible to leave the final result persisted, yes.

Commit:
SPARK-16478 (changed way sccGraph is persisted and unpersist done with one helper variable)

wesolowskim commented:
I agree it is possible however quite hard in this case. You loose reference to workGraph few times before materialization occurs, and then few graphs get materialized at the same time. Testing is also troublesome, because you have to actually check if something didn't get unpersisted too early and that requires big graph (and long running program) or sophisticated code to check it. 
I pushed my last commit that corrects mistakes you pointed out and additionally removes some cache() that was superfluous.

Commit:
SPARK-16478 (removed unnecessary cache and if statement)

srowen commented on graphx/src/main/scala/org/apache/spark/graphx/lib/StronglyConnectedComponents.scala:
```diff
@@ -106,6 +116,16 @@ object StronglyConnectedComponents {
+        sccGraph.vertices.count()
+        sccGraph.edges.count()
+        // sccGraph materialized so, unpersist can be done on previous
+        if(prevSccGraph != sccGraph)
```
I think this will fail style checks but let's see.
Hm, I think it's straightforward to follow the same pattern for the other RDD, but yes it's updated in enough places that it's a fair bit of extra code. It might be possible to refactor this nicely in some method.
Not essential, maybe. This code isn't used much anyway. A partial improvement may be a sufficient win.

wesolowskim replied:
I forgot to push one commit. Nor If statement neither cache() were necessary.

srowen commented on graphx/src/main/scala/org/apache/spark/graphx/lib/StronglyConnectedComponents.scala:
```diff
@@ -106,6 +116,11 @@ object StronglyConnectedComponents {
         },
         (final1, final2) => final1 || final2)
+
+        // ensure sccGraph's rdd are marked as recently used to not be evicted between iterations
```
OK, the rest of the changes much looks like an improvement, at the least. I am not as sure why this is necessary. It's already materialized at this point.

Commit:
SPARK-16478 (removed unnecessary count on sccGraph)

Commit:
SPARK-16478 (additional optimization - skip pregel if computing cannot change returned graph)

wesolowskim commented:
I removed counts at the end of outside loop. I added it before because without it I still encountered problems, but I guess something else must have been wrong. I reasoned that despite the fact that sccGraph is persisted it can be evicted before it is used again. It was quite strange, because it doesn't happen much between sccGraph reuse and this additional counts. I tested it once again and it works with current version. 
I added change that saves some computing time in some cases - if while loop leaves because iter>maxIter Pregels are useless.

srowen commented:
I don't think you can reason about whether something's evicted as it's up to the runtime. Here the RDD has to be materialized before the method returns because its predecessors will have been unpersisted, but that is already done in the loop.

What's the reason the final loop iteration doesn't need to compute this? it wasn't obvious to me; a comment would be useful for future readers.

Commit:
SPARK-16478 (added additional comment explaining latest if statement)

wesolowskim commented:
Added comment explaining if.

wesolowskim commented:
Is it sufficient right now or should I do something more?

srowen commented:
Jenkins retest this please

srowen commented:
LGTM, will leave open for a bit for comments

SparkQA commented:
**[Test build #62464 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62464/consoleFull)** for PR 14137 at commit [`84a8363`](https://github.com/apache/spark/commit/84a8363adfac60557a65d91216a5f790df5961da).

SparkQA commented:
**[Test build #62464 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62464/consoleFull)** for PR 14137 at commit [`84a8363`](https://github.com/apache/spark/commit/84a8363adfac60557a65d91216a5f790df5961da).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62464/
Test PASSed.
